pkg/k8s/serviceAccountHelper.go

```diff
@@ -0,0 +1,60 @@
+package k8s
+
+import (
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type serviceAccountHelper struct {
+	factory ClientFactory
+	cache   *v1.ServiceAccount
+}
```

Comment:
This is the same "wrapper" design as `ServiceAccountWrap`, which I consider bad design.

I imagine something like this:

```go
// ServiceAccountTokenSecretRetriever retrieves the service account
// token secret for a service account.
//
// For illustration it has multiple methods but each is just a variant of
// invocation of the same functionality.
type ServiceAccountTokenSecretRetriever interface {
    // ForServiceAccount retrieves ... for the given service account object
    // using the K8s client in `ctx`.
    ForObj(Context ctx, serviceAccount *v1.ServiceAccount) (v1.Secret, error)

    // ForName retries ... for the named service account
    // using the K8s client in `ctx`.
    ForName(Context ctx, String serviceAccountName, String namespace) (v1.Secret, error)
}

var ctxKey = ServiceAccountTokenSecretRetriever // use type as key

// GetServiceAccountTokenSecretRetrieverFromContext provides the
// `ServiceAccountTokenSecretRetriever` instance from the given context,
// or nil if it doesn't contain one.
func GetServiceAccountTokenSecretRetrieverFromContext(Context ctx) *ServiceAccountTokenSecretRetriever {
    return ctx.Value(ctxKey).(ServiceAccountTokenSecretRetriever)
}

// GetOrCreateServiceAccountTokenSecretRetriever provides the
// `ServiceAccountTokenSecretRetriever` from the given context,
// or creates a new instance if none exists in the context.
// The returned context either is or extends the given context and
// has a `ServiceAccountTokenSecretRetriever` instance stored.
func GetOrCreateServiceAccountTokenSecretRetrieverFromContext(Context ctx) (Context, *ServiceAccountTokenSecretRetriever) {
    instance := GetServiceAccountTokenSecretRetrieverFromContext(ctx)
    if instance != nil {
        ctx = context.WithValue(ctx, ctxKey, serviceAccountTokenSecretRetrieverImpl{})
    instance := GetServiceAccountTokenSecretRetrieverFromContext(ctx)
    }
    return ctx, instance
}

// --- default impl ---

type serviceAccountTokenSecretRetrieverImpl struct {}

func (r *serviceAccountTokenSecretRetrieverImpl) ForObj(...) (...) {
    ...
}

func (r *serviceAccountTokenSecretRetrieverImpl) ForName(...) (...) {
    ...
}
```

A consumer looks like this:

```go
func consumerFunc(Context ctx) {
  saObj := ...
  subCtx := ... // e.g. adding K8y client
  _, retrieveSaTokenSec := GetServiceAccountTokenSecretRetrieverFromContext(ctx)
  sec, err := retrieveSaTokenSec.FromObj(subCtx, saObj)
  ...
}
```

Reply:
As learned recently, `Context` should not be used in that way. The main reason is that types no longer have a formal dependency declaration and it gets hard to understand what other object an object is actually using.

So consider my proposal as invalid.